File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-doc-5-4/BestFitTrade.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class BestFitTrade : ISerialized
    {
        public BestFitTrade(XmlNode xmlNode)
        {
            XmlNode tradeIdentifierNode = xmlNode.SelectSingleNode("tradeIdentifier");
            
            if (tradeIdentifierNode != null)
            {
                if (tradeIdentifierNode.Attributes["href"] != null || tradeIdentifierNode.Attributes["id"] != null) 
                {
                    if (tradeIdentifierNode.Attributes["id"] != null) 
                    {
                        tradeIdentifierIDRef_ = tradeIdentifierNode.Attributes["id"].Value;
                        TradeIdentifier ob = new TradeIdentifier(tradeIdentifierNode);
                        IDManager.SetID(tradeIdentifierIDRef_, ob);
                    }
                    else if (tradeIdentifierNode.Attributes["href"] != null)
                    {
                        tradeIdentifierIDRef_ = tradeIdentifierNode.Attributes["href"].Value;
                    }
                    else
                    {
                        tradeIdentifier_ = new TradeIdentifier(tradeIdentifierNode);
                    }
                }
                else
                {
                    tradeIdentifier_ = new TradeIdentifier(tradeIdentifierNode);
                }
            }
            
        
            XmlNodeList differencesNodeList = xmlNode.SelectNodes("differences");
            
            if (differencesNodeList != null)
            {
                this.differences_ = new List<TradeDifference>();
                foreach (XmlNode item in differencesNodeList)
                {
                    if (item.Attributes["href"] != null || item.Attributes["id"] != null) 
                    {
                        if (item.Attributes["id"] != null) 
                        {
                            differencesIDRef_ = item.Attributes["id"].Value;
                            differences_.Add(new TradeDifference(item));
                            IDManager.SetID(differencesIDRef_, differences_[differences_.Count - 1 ]);
                        }
                        else if (item.Attributes["href"] != null)
                        {
                            differencesIDRef_ = item.Attributes["href"].Value;
                        }
                        else
                        {
                        differences_.Add(new TradeDifference(item));
                        }
                    }
                    else
                    {
                        differences_.Add(new TradeDifference(item));
                    }
                }
            }
            
        
        }
        
    
        #region TradeIdentifier_
        private TradeIdentifier tradeIdentifier_;
        public TradeIdentifier TradeIdentifier_
        {
            get
            {
                if (this.tradeIdentifier_ != null)
                {
                    return this.tradeIdentifier_; 
                }
                else if (this.tradeIdentifierIDRef_ != null)
                {
                    tradeIdentifier_ = IDManager.getID(tradeIdentifierIDRef_) as TradeIdentifier;
                    return this.tradeIdentifier_; 
                }
                else
                {
                      return this.tradeIdentifier_; 
                }
            }
            set
            {
                if (this.tradeIdentifier_ != value)
                {
                    this.tradeIdentifier_ = value;
                }
            }
        }
        #endregion
        
        public string tradeIdentifierIDRef_ { get; set; }
        #region Differences_
        private List<TradeDifference> differences_;
        public List<TradeDifference> Differences_
        {
            get
            {
                if (this.differences_ != null)
                {
                    return this.differences_; 
                }
                else if (this.differencesIDRef_ != null)
                {
                    return this.differences_; 
                }
                else
                {
                      return this.differences_; 
                }
            }
            set
            {
                if (this.differences_ != value)
                {
                    this.differences_ = value;
                }
            }
        }
        #endregion
        
        public string differencesIDRef_ { get; set; }
        
    
        
    
    }
    
}

File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-doc-5-4/BestFitTrade.cpp
// BestFitTrade.cpp 
#include "BestFitTrade.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

BestFitTrade::BestFitTrade(TiXmlNode* xmlNode)
: ISerialized(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //tradeIdentifierNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* tradeIdentifierNode = xmlNode->FirstChildElement("tradeIdentifier");

   if(tradeIdentifierNode){tradeIdentifierIsNull_ = false;}
   else{tradeIdentifierIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- tradeIdentifierNode , address : " << tradeIdentifierNode << std::endl;
   #endif
   if(tradeIdentifierNode)
   {
      if(tradeIdentifierNode->Attribute("href") || tradeIdentifierNode->Attribute("id"))
      {
          if(tradeIdentifierNode->Attribute("id"))
          {
             tradeIdentifierIDRef_ = tradeIdentifierNode->Attribute("id");
             tradeIdentifier_ = boost::shared_ptr<TradeIdentifier>(new TradeIdentifier(tradeIdentifierNode));
             tradeIdentifier_->setID(tradeIdentifierIDRef_);
             IDManager::instance().setID(tradeIdentifierIDRef_,tradeIdentifier_);
          }
          else if(tradeIdentifierNode->Attribute("href")) { tradeIdentifierIDRef_ = tradeIdentifierNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { tradeIdentifier_ = boost::shared_ptr<TradeIdentifier>(new TradeIdentifier(tradeIdentifierNode));}
   }

   //differencesNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* differencesNode = xmlNode->FirstChildElement("differences");

   if(differencesNode){differencesIsNull_ = false;}
   else{differencesIsNull_ = true;}

   if(differencesNode)
   {
      for(differencesNode; differencesNode; differencesNode = differencesNode->NextSiblingElement("differences")){
          #ifdef ConsolePrint
              FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- differencesNode , address : " << differencesNode << std::endl;
          #endif
          if(differencesNode->Attribute("href") || differencesNode->Attribute("id"))
          {
              if(differencesNode->Attribute("id"))
              {
                  differencesIDRef_ = differencesNode->Attribute("id");
                  differences_.push_back(boost::shared_ptr<TradeDifference>(new TradeDifference(differencesNode)));
                  differences_.back()->setID(differencesIDRef_);
                  IDManager::instance().setID(differencesIDRef_, differences_.back());
              }
              else if(differencesNode->Attribute("href")) { differencesIDRef_ = differencesNode->Attribute("href");}
              else { QL_FAIL("id or href error"); }
          }
          else { differences_.push_back(boost::shared_ptr<TradeDifference>(new TradeDifference(differencesNode)));}
      }
   }
   else {
       #ifdef ConsolePrint
           FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- differencesNode , address : " << differencesNode << std::endl;
       #endif
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<TradeIdentifier> BestFitTrade::getTradeIdentifier()
{
   if(!this->tradeIdentifierIsNull_){
        if(tradeIdentifierIDRef_ != ""){
             return boost::shared_static_cast<TradeIdentifier>(IDManager::instance().getID(tradeIdentifierIDRef_));
        }else{
             return this->tradeIdentifier_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<TradeIdentifier>();
   }
}
std::vector<boost::shared_ptr<TradeDifference>> BestFitTrade::getDifferences()
{
   if(!this->differencesIsNull_){
        if(differencesIDRef_ != ""){
             return this->differences_;
        }else{
             return this->differences_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return std::vector<boost::shared_ptr<TradeDifference>>();
   }
}
}

